crates/ty_python_semantic/src/types.rs

```diff
@@ -4547,6 +4548,36 @@ impl<'db> Type<'db> {
             .to_instance(db)
             .expect("type should be convertible to instance type");
 
+        // Check if __new__ explicitly returns a different type than the instance
```

Comment:
If it does, we shouldn't attempt to call `__init__` at all. So this check should come before the `__init__` call, and we should bypass the `__init__` call if we find a return type that isn't a subclass of the self type.